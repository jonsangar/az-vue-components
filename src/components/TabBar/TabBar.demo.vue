<template lang="pug">
  .tabBarDemo
    Demo(title="Uso Básico")
      az-tabBar(v-model="active" :fixed="false")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap") Tab
        az-tabBar-item(icon="desktop") Tab
        az-tabBar-item(icon="large") Tab

    Demo(title="Seleccionar por nombre")
      az-tabBar(v-model="name" :fixed="false")
        az-tabBar-item(icon="palm" name="palm")
        az-tabBar-item(icon="lap" name="lap")
        az-tabBar-item(icon="desktop" name="desktop")
        az-tabBar-item(icon="large" name="large")

    Demo(title="Barra de navegación")
      az-tabBar(v-model="route" :fixed="false")
        az-tabBar-item(icon="palm" to="home") Home
        az-tabBar-item(icon="lap" to="shop") Shop
        az-tabBar-item(icon="desktop" to="faqs") Faqs
        az-tabBar-item(icon="large" to="contact") Contact

    Demo(title="Mostrar etiqueta")
      az-tabBar(v-model="badge" :fixed="false")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap" dot) Tab
        az-tabBar-item(icon="desktop" info="5") Tab
        az-tabBar-item(icon="large" info="10") Tab

    Demo(title="Mostrar etiqueta")
      az-tabBar(v-model="custom" :fixed="false" active-color="var(--color-brand)" inactive-color="var(--color-gray1)")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap" dot) Tab
        az-tabBar-item(icon="desktop" info="5") Tab
        az-tabBar-item(icon="large" info="10") Tab
</template>

<script>
export default {
  data() {
    return {
      active: 0,
      name: 'palm',
      route: 0,
      badge: 0,
      custom: 0,
    };
  },
};
</script>

<style lang="scss">
.tabBarDemo {
  .az-tabBar {
    margin-bottom: simple();
    background-color: color(gray7);
  }
}
</style>
