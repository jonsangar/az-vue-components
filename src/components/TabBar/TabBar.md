---
device: true
---

# TabBar

### Instalación

```javascript
import { TabBar } from 'didor';

Vue.use(TabBar);
```

---

### Uso

#### Uso básico

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm">Tab</az-tabBar-item>
  <az-tabBar-item icon="lap">Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop">Tab</az-tabBar-item>
  <az-tabBar-item icon="large">Tab</az-tabBar-item>
</az-tabBar>
```

``` js
export default {
  data() {
    return {
      active: 0
    }
  }
}
```

#### Seleccionar por nombre

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm" name="palm"/>
  <az-tabBar-item icon="lap" name="lap"/>
  <az-tabBar-item icon="desktop" name="desktop"/>
  <az-tabBar-item icon="large" name="large"/>
</az-tabBar>
```

``` js
export default {
  data() {
    return {
      active: 0
    }
  }
}
```

#### Barra de navegación

Cuando funciona como barra de navegación, no es necesario indicar el item activo, lo toma directamente de la ruta de navegación.

Debemos pasar el nombre de la ruta o un objeto si queremos pasar parámetros:

``` html
<az-tabBar>
  <az-tabBar-item icon="palm" to="home">home</az-tabBar-item>
  <az-tabBar-item icon="lap" :to="{name: 'shop', params: {category: 'phones'}}">Shop</az-tabBar-item>
  <az-tabBar-item icon="desktop" to="faqs">Faqs</az-tabBar-item>
  <az-tabBar-item icon="large" to="contact">Contact</az-tabBar-item>
</az-tabBar>
```

#### Mostrar etiqueta

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm">Tab</az-tabBar-item>
  <az-tabBar-item icon="lap" dot>Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop" info="5">Tab</az-tabBar-item>
  <az-tabBar-item icon="large" info="10">Tab</az-tabBar-item>
</az-tabBar>
```

#### Colores personalizados

``` html
<az-tabBar v-model="active" active-color="var(--color-brand)" inactive-color="var(--color-gray1)">
  <az-tabBar-item icon="palm"/>
>Tab</az-tabBar-item>
  <az-tabBar-item icon="lap" dot>Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop" info="5">Tab</az-tabBar-item>
  <az-tabBar-item icon="large" info="10">Tab</az-tabBar-item>
</az-tabBar>
```

---

### API TabBar

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| v-model | Identifica la pestaña activa | `String` `Number` | `0` |
| fixed | Mantiene la barra fija en la parte de abajo de la pantalla | `Boolean` | `true` |
| safeArea | Habilita un espacio extra en la parte inferior de la barra para dispositivos con un botón en la parte inferior de la pantalla | `Boolean` | `true` |
| z-index | Permite especificar el valor del z-index de la barra | `Number` | `1` |
| active-Color | Color de un item activo | `String` | -- |
| inactive-Color | Color de un item inactivo | `String` | -- |

### Eventos TabBar

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| change | Se lanza cuando cambia el valor del TabBar | **active**: Index de la pestaña activa         |


### API TabBarItem

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| icon | Nombre del icono | `String` | -- |
| name | Identificador de la pestaña | `String` `Number` | -- |
| dot | Muestra un punto rojo sobre el icono | `Boolean` | `false` |
| info | Muestra un mensaje en rojo sobre el icono | `String` | -- |
| to | Nombre de la ruta utilizando vue-router | `String` `Object` | -- |

### Eventos TabBarItem

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| click | Se lanza cuando hacemos click en un item del TabBar | -         |
