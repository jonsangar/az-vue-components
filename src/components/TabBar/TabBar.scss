$tabBar-height: double();
$tabBar-background-color: #ffffff;

.az-tabBar {
  $rootClass: &;

  display: flex;
  width: 100%;
  height: $tabBar-height;
  background-color: $tabBar-background-color;

  &--fixed {
    position: fixed;
    bottom: 0;
    left: 0;
  }

  &--safe-area {
    padding-bottom: constant(safe-area-inset-bottom);
    padding-bottom: env(safe-area-inset-bottom);
    padding-bottom: var(--notch-inset-bottom);
  }
}
