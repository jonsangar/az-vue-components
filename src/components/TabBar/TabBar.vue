<template lang="pug">
  .az-tabBar(:class="tabBarClass" :style="tabBarStyle" @keyup="onKeyup")
    slot
</template>

<script>
export default {
  props: {
    value: {
      type: [Number, String],
      default: 0,
    },
    activeColor: {
      type: String,
      required: false,
    },
    inactiveColor: {
      type: String,
      required: false,
    },
    fixed: {
      type: Boolean,
      default: true,
    },
    zIndex: {
      type: Number,
      default: 1,
    },
    safeArea: {
      type: Boolean,
      default: true,
    },
  },

  watch: {
    value: 'setActiveItem',
  },

  computed: {
    tabBarClass() {
      return [this.fixed ? 'az-tabBar--fixed' : '', this.safeArea ? 'az-tabBar--safeArea' : ''];
    },

    tabBarStyle() {
      return {
        'z-index': this.zIndex,
      };
    },
  },

  methods: {
    setActiveItem() {
      this.$children.forEach((item, index) => {
        if (item.to) return;
        item.active = (item.name || index) === this.value;
      });
    },

    onKeyup(e) {
      this.$emit('keyup',e);
    },

    onChange(active) {
      if (active !== this.value) {
        this.$emit('input', active);
        this.$emit('change', active);
      }
    },
  },

  mounted() {
    this.setActiveItem();
  },
};
</script>

<style src="./TabBar.scss" lang="scss"></style>
