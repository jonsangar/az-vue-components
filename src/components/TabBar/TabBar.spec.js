import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import TabBar from './TabBar.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('TabBar.vue', () => {

  const build = () => {
    const wrapper = shallowMount(TabBar, {
      localVue,
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente TabBar con las props por defecto con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.value).toBe(0)
    expect(wrapper.vm.fixed).toBe(true)
    expect(wrapper.vm.safeArea).toBe(true)
    expect(wrapper.vm.zIndex).toBe(1)
  })
})

describe('TabBar.vue', () => {

  const build = () => {
    const wrapper = shallowMount(TabBar, {
      localVue,
      propsData: {
        value: 1,
        activeColor: 'brand',
        inactiveColor: 'light',
        fixed: false,
        safeArea: false,
        zIndex: 10,
      },
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que se actualizan las props cuando se asignan las props', async () => {
    const { wrapper } = build()

    expect(wrapper.vm.value).toBe(1)
    expect(wrapper.vm.activeColor).toBe('brand')
    expect(wrapper.vm.inactiveColor).toBe('light')
    expect(wrapper.vm.fixed).toBe(false)
    expect(wrapper.vm.safeArea).toBe(false)
    expect(wrapper.vm.zIndex).toBe(10)
  });
})

describe('TabBar.vue', () => {
  const e = {
    event: {
      type: 'change'
    }
  }

  const build = () => {
    const wrapper = shallowMount(TabBar, {
      localVue,
      propsData: {
        name: 'MyTitle',
        itemID: 'MyItem',
        settingsActive: true,
      },
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que emite el evento keyup cuando se teclea en la pestaña', async () => {
    const { wrapper } = build()

    await wrapper.vm.onKeyup(e)
    expect(wrapper.emitted().keyup).toBeTruthy()
    expect(wrapper.emitted().keyup.length).toBe(1)
    expect(wrapper.emitted().keyup[0]).toEqual([e])
  });

  it('Comprueba que emite el evento input cuando se cambia de pestaña', async () => {
    const { wrapper } = build()

    await wrapper.vm.onChange(e)
    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().input.length).toBe(1)
    expect(wrapper.emitted().input[0]).toEqual([e])
  });

  it('Comprueba que emite el evento change cuando se cambia de pestaña', async () => {
    const { wrapper } = build()

    await wrapper.vm.onChange(e)
    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.emitted().change.length).toBe(1)
    expect(wrapper.emitted().change[0]).toEqual([e])
  });
})
