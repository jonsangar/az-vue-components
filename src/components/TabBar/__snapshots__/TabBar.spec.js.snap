// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar.vue Comprueba renderizado del componente TabBar con las props por defecto con snapshot 1`] = `<div class="az-tabBar az-tabBar--fixed az-tabBar--safeArea" style="z-index: 1;"></div>`;
